Wrappers/Python/ccpi/viewer/utils/hdf5_io.py

```diff
@@ -55,64 +57,106 @@ def __init__(self):
+            shape = np.shape(f[self._DatasetName])
             # print("keys:", list(f.keys()))
             # print(shape)
             ue = info.Get(vtk.vtkStreamingDemandDrivenPipeline.UPDATE_EXTENT())
             # Note that we flip the update extents because VTK is Fortran order
             # whereas h5py reads in C order. When writing we pretend that the
             # data was C order so we have to flip the extents/dimensions.
             if len(shape) == 3:
-                data = f[self.__DatasetName][ue[4]:ue[5]+1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
+                data = f[self._DatasetName][ue[4]:ue[5]+1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
             elif len(shape) == 4:
-                data = f[self.__DatasetName][self.__4DIndex][
-                    ue[4]:ue[5] + 1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
+                if self._4DIndex == 0:
+                    data = f[self._DatasetName][self._4DSliceIndex, ue[4]:ue[5]+1, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
+                elif self._4DIndex == 1:
+                    data = f[self._DatasetName][ue[4]:ue[5]+1, self._4DSliceIndex, ue[2]:ue[3]+1, ue[0]:ue[1]+1]
+                elif self._4DIndex == 2:
+                    data = f[self._DatasetName][ue[4]:ue[5]+1,  ue[2]:ue[3]+1, self._4DSliceIndex, ue[0]:ue[1]+1]
+                elif self._4DIndex == 3:
+                    data = f[self._DatasetName][ue[4]:ue[5]+1,  ue[2]:ue[3]+1,  ue[0]:ue[1]+1, self._4DSliceIndex]
+            else:
+                raise Exception("Currently only 3D and 4D datasets are supported.")
             # print("attributes: ", f.attrs.items())
             output = dsa.WrapDataObject(vtk.vtkImageData.GetData(outInfo))
             output.SetExtent(ue)
-            output.PointData.append(data.ravel(), self.__DatasetName)
-            output.PointData.SetActiveScalars(self.__DatasetName)
-            return 1
+            output.PointData.append(data.ravel(), self._DatasetName)
+            output.PointData.SetActiveScalars(self._DatasetName)
+            return output
+        
 
     def SetFileName(self, fname):
-        if fname != self.__FileName:
+        if fname != self._FileName:
             self.Modified()
-            self.__FileName = fname
+            if self._DatasetName is not None:
+                with h5py.File(fname, 'r') as f:
+                    if not (self._DatasetName in f):
+                        raise Exception("No dataset named {} exists in {}.".format(self._DatasetName, fname))
+            self._FileName = fname
 
     def GetFileName(self):
-        return self.__FileName
+        return self._FileName
 
     def SetDatasetName(self, lname):
-        if lname != self.__DatasetName:
+        if lname != self._DatasetName:
             self.Modified()
-            self.__DatasetName = lname
+            if self._FileName is not None:
+                with h5py.File(self._FileName, 'r') as f:
+                    if not (lname in f):
```

Comment:
does this look in the `key`s?

Reply:
It looks at the keys but also works if it's a path such as entry1/tomo_entry/data etc
https://docs.h5py.org/en/stable/high/group.html#h5py.Group.__contains__